src/replay/consensus/core.zig

```diff
@@ -878,6 +974,407 @@ fn pushVote(
     }
 }
 
+/// Returns an owned slice of tpu vote sockets for the leaders of the next N fanout
+/// slots. Leaders and sockets are deduped. Caller owns the returned slice.
+fn upcomingLeaderTpuVoteSockets(
+    allocator: Allocator,
+    current_slot: Slot,
+    leader_schedule: *sig.core.leader_schedule.LeaderScheduleCache,
+    gossip_table_rw: *sig.sync.RwMux(sig.gossip.GossipTable),
+    fanout_slots: u64,
+) ![]SocketAddr {
+    var seen_leaders: std.AutoHashMapUnmanaged(Pubkey, void) = .empty;
+    defer seen_leaders.deinit(allocator);
+
+    for (0..fanout_slots) |n_slots| {
+        const target_slot = current_slot + n_slots;
+        if (leader_schedule.slotLeader(target_slot)) |leader| {
+            try seen_leaders.put(allocator, leader, {});
+        }
+    }
+
+    const gossip_table, var gossip_table_lg = gossip_table_rw.readWithLock();
+    defer gossip_table_lg.unlock();
+
+    var seen_sockets: std.AutoHashMapUnmanaged(SocketAddr, void) = .empty;
+    defer seen_sockets.deinit(allocator);
+
+    var leader_sockets: std.ArrayListUnmanaged(SocketAddr) = .empty;
+    defer leader_sockets.deinit(allocator);
+
+    var leader_iter = seen_leaders.keyIterator();
+    while (leader_iter.next()) |leader_pubkey_ptr| {
+        const leader_pubkey = leader_pubkey_ptr.*;
+        const contact_info = gossip_table.getThreadSafeContactInfo(leader_pubkey) orelse continue;
+
+        const socket_addr = contact_info.tpu_addr orelse continue;
+
+        const gop = try seen_sockets.getOrPut(allocator, socket_addr);
+        if (!gop.found_existing) {
+            try leader_sockets.append(allocator, socket_addr);
+        }
+    }
+
+    return leader_sockets.toOwnedSlice(allocator);
+}
+
+fn sendVoteTransaction(
+    logger: Logger,
+    allocator: Allocator,
+    vote_tx: Transaction,
+    tpu_address: SocketAddr,
+) !void {
+    var serialized: std.ArrayListUnmanaged(u8) = .empty;
+    defer serialized.deinit(allocator);
+
+    try sig.bincode.write(serialized.writer(allocator), vote_tx, .{});
+
+    const socket = switch (tpu_address) {
+        .V4 => network.Socket.create(.ipv4, .udp) catch |err| {
+            logger.err().logf("Failed to create socket for vote send: {}", .{err});
+            return err;
+        },
+        .V6 => network.Socket.create(.ipv6, .udp) catch |err| {
+            logger.err().logf("Failed to create socket for vote send: {}", .{err});
+            return err;
+        },
+    };
+    defer socket.close();
```

Comment:
```suggestion
    const addr_family: network.AddressFamily = switch (tpu_address) {
        .V4 => .ipv4,
        .V6 => .ipv6,
    };
    const socket = network.Socket.create(addr_family, .udp) catch |err| {
        logger.err().logf("Failed to create socket for vote send: {}", .{err});
        return err;
    };
    defer socket.close();
```
Though I think this is the wrong place to do this; I think a pair of ipv4 and ipv6 UDP sockets should be created outside of the loop and passed as parameters to this function, with the one appropriate for the tpu_address used.